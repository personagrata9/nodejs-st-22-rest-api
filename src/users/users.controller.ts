import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  Param,
  ParseUUIDPipe,
  Post,
  Put,
  Query,
} from '@nestjs/common';
import { CreateUserDto } from './dto/create-user.dto';
import { IUser } from './models/user.model';
import { UsersService } from './services/users.service';
import { UpdateUserDto } from './dto/update-user.dto';
import { UserByIdPipe } from './validation/pipes/user-by-id.pipe';

@Controller('v1/users')
export class UsersController {
  constructor(private readonly usersService: UsersService) {}

  @Get()
  async getAutoSuggestUsers(@Query() query: any) {
    const { limit = 10, offset = 0, loginSubstring } = query;
    return {
      items: await this.usersService.findAutoSuggestUsers(
        +limit,
        +offset,
        loginSubstring,
      ),
      limit: +limit,
      offset: +offset,
      count: await this.usersService.count(),
    };
  }

  @Get(':id')
  async getById(
    @Param('id', new ParseUUIDPipe({ version: '4' }), UserByIdPipe) id: string,
  ): Promise<IUser> {
    return await this.usersService.findOneById(id);
  }

  @Post()
  async create(@Body() createUserDto: CreateUserDto): Promise<IUser> {
    return await this.usersService.create(createUserDto);
  }

  @Put(':id')
  async update(
    @Param('id', new ParseUUIDPipe({ version: '4' }), UserByIdPipe) id: string,
    @Body() updateUserDto: UpdateUserDto,
  ): Promise<IUser> {
    return await this.usersService.update(id, updateUserDto);
  }

  @Delete(':id')
  @HttpCode(204)
  async delete(
    @Param('id', new ParseUUIDPipe({ version: '4' }), UserByIdPipe) id: string,
  ) {
    return await this.usersService.delete(id);
  }
}
