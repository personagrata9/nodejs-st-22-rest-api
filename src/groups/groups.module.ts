import 'dotenv/config';
import { Module } from '@nestjs/common';
import { SequelizeModule } from '@nestjs/sequelize';
import { Group } from './models/group.model';
import { GroupsService } from './services/groups.service';
import { InMemoryGroupsRepository } from './repository/in-memory-groups.repository';
import { SequelizeGroupsRepository } from './repository/sequelize-groups.repository';
import { GroupsController } from './controllers/groups.controller';
import { UserGroup } from './models/user-group.model.';
import { UsersModule } from '../users/users.module';

@Module({
  imports: [SequelizeModule.forFeature([Group, UserGroup]), UsersModule],
  controllers: [GroupsController],
  providers: [
    GroupsService,
    {
      provide: 'GroupsRepository',
      useClass:
        process.env.NODE_ENV === 'test'
          ? InMemoryGroupsRepository
          : SequelizeGroupsRepository,
    },
  ],
})
export class GroupsModule {}
