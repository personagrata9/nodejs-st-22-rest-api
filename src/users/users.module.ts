import 'dotenv/config';
import { Module } from '@nestjs/common';
import { InMemoryUsersRepository } from './repository/in-memory-users.repository';
import { SequelizeUsersRepository } from './repository/sequelize-users.repository';
import { UsersService } from './services/users.service';
import { UsersController } from './users.controller';

@Module({
  controllers: [UsersController],
  providers: [
    UsersService,
    {
      provide: 'UsersRepository',
      useClass:
        process.env.NODE_ENV === 'test'
          ? InMemoryUsersRepository
          : SequelizeUsersRepository,
    },
  ],
})
export class UsersModule {}
