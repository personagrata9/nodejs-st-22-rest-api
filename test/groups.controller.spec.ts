import { Test, TestingModule } from '@nestjs/testing';
import { GroupsController } from '../src/groups/controllers/groups.controller';
import { GroupsService } from '../src/groups/services/groups.service';
import { GroupsServiceMock } from './__mocks__/groups.service';
import { UsersService } from '../src/users/services/users.service';
import { UsersServiceMock } from './__mocks__/users.service.mock';
import { IGroup } from '../src/groups/interfaces/group.interface';
import { IPaginatedItemsResult } from '../src/common/interfaces/paginated-items-result.interface';
import { group } from './fixtures/group.fixture';
import { createGroupDto } from './fixtures/create-group-dto.fixture';
import { NotUniqueError } from '../src/common/errors/not-unique.error';
import { BadRequestException } from '@nestjs/common';
import { updateGroupDto } from './fixtures/update-group-dto.fixture';
import { ISuccessResponse } from 'src/common/interfaces/success-response.interface';
import { user } from './fixtures/user.fixture';

describe('GroupsController', () => {
  let controller: GroupsController;
  let service: any;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [GroupsController],
      providers: [
        {
          provide: GroupsService,
          useClass: GroupsServiceMock,
        },
        {
          provide: UsersService,
          useClass: UsersServiceMock,
        },
      ],
    }).compile();

    controller = module.get<GroupsController>(GroupsController);
    service = module.get<GroupsService>(GroupsService);
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
  });

  it('should return array of groups', async () => {
    const limit = 10;
    const offset = 0;

    service.findAll.mockResolvedValue({
      items: [group],
      limit,
      offset,
      count: 1,
    });

    const result: IPaginatedItemsResult<IGroup> = await controller.getAll({
      limit,
      offset,
    });

    expect(result).toEqual({ items: [group], limit, offset, count: 1 });
    expect(service.findAll.mock.calls[0]).toMatchSnapshot();
  });

  it('should return group', async () => {
    service.findOneById.mockResolvedValue(group);

    const result: IGroup = await controller.getById(group);

    expect(result).toEqual(group);
  });

  it('should create a group', async () => {
    service.create.mockResolvedValue(group);

    const result: IGroup = await controller.create(createGroupDto);

    expect(result).toEqual(group);
  });

  it('should throw an error when creating a group with not unique name', async () => {
    service.create.mockRejectedValue(
      new NotUniqueError('group', 'name', createGroupDto.name),
    );

    try {
      await controller.create(createGroupDto);
    } catch (error) {
      expect(error).toBeInstanceOf(BadRequestException);
      expect(error.message).toBe(
        `group with name ${createGroupDto.name} already exists, please choose another name`,
      );
    }
  });

  it('should add users to group', async () => {
    service.addUsersToGroup.mockResolvedValue();

    const result: ISuccessResponse = await controller.addUsersToGroup(group, {
      userIds: [user.id],
    });

    expect(result.message).toBe(
      `users were successfully added to group with id ${group.id}`,
    );
  });

  it('should update existing group', async () => {
    const updatedGroup: IGroup = { ...group, ...updateGroupDto };

    service.update.mockResolvedValue(updatedGroup);

    const result: IGroup = await controller.update(group, updateGroupDto);

    expect(result).toEqual(updatedGroup);
  });

  it('should throw an error when updating a group with not unique name', async () => {
    service.update.mockRejectedValue(
      new NotUniqueError('group', 'name', updateGroupDto.name),
    );

    try {
      await controller.update(group, updateGroupDto);
    } catch (error) {
      expect(error).toBeInstanceOf(BadRequestException);
      expect(error.message).toBe(
        `group with name ${updateGroupDto.name} already exists, please choose another name`,
      );
    }
  });

  it('should delete existing group', async () => {
    const result = await controller.delete(group);
    expect(result).toBeUndefined();
  });
});
