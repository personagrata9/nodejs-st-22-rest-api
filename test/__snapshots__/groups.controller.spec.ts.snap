// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupsController should return array of groups 1`] = `
Array [
  10,
  0,
]
`;
